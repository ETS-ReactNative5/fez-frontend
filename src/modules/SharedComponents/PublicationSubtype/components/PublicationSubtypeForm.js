import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { loadPublicationSubtypesList } from 'actions';

import MenuItem from 'material-ui/MenuItem';
import SelectField from 'material-ui/SelectField';

export class PublicationSubtypeForm extends Component {
    static propTypes = {
        onChange: PropTypes.func,
        locale: PropTypes.object,
        subtypesList: PropTypes.array,
        selectedValue: PropTypes.string,
        dataSourceConfig: PropTypes.object,
        dispatch: PropTypes.func,
        vocabId: PropTypes.number,
        className: PropTypes.string
    };

    static defaultProps = {
        dataSourceConfig: {
            text: 'controlled_vocab.cvo_title',
            value: 'controlled_vocab.cvo_title'
        },
        locale: {
            label: 'Publication subtype'
        }
    };

    constructor(props) {
        super(props);
        this.state = {
            selectedValue: props.selectedValue || null
        };
    }

    componentDidMount() {
        this.props.dispatch(loadPublicationSubtypesList(this.props.vocabId));
    }

    componentWillUpdate(nextProps, nextState) {
        if (this.props.onChange && nextState.selectedValue !== this.state.selectedValue) this.props.onChange(nextState.selectedValue);
    }

    _onSubtypeSelected = (event, index, value) => {
        this.setState({
            selectedValue: value
        });
    };

    getValue = (item, path) => {
        return path.split('.').reduce((objectValue, pathProperty) => objectValue[pathProperty], item);
    };

    render() {
        const { locale, subtypesList, dataSourceConfig } = this.props;
        const renderSubTypeItems = subtypesList.map((item) => {
            const value = this.getValue(item, dataSourceConfig.value);
            const text = this.getValue(item, dataSourceConfig.text);
            return <MenuItem value={ value } primaryText={ text } key={ value }/>;
        });

        return (
            <SelectField
                name="selectedValue"
                fullWidth
                className={ this.props.className }
                value={ this.state.selectedValue }
                maxHeight={ 250 }
                onChange={ this._onSubtypeSelected }
                floatingLabelText={ locale.label }>
                <MenuItem
                    primaryText={ locale.label }
                    disabled/>
                { renderSubTypeItems }
            </SelectField>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        subtypesList: state.get('publicationSubtypesReducer').subtypesList || []
    };
};

export default connect(mapStateToProps)(PublicationSubtypeForm);
