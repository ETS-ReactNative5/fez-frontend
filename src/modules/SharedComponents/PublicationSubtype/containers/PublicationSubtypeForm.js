import React, { Component } from 'react';
import PropTypes from 'prop-types';

import {connect} from 'react-redux';
import { loadPublicationSubtypesList } from 'actions';

import MenuItem from 'material-ui/MenuItem';
import SelectField from 'material-ui/SelectField';
import PublicationSubtypeForm from '../components/PublicationSubtypeForm';

class PublicationSubtypeFormContainer extends Component {
    static propTypes = {
        vocabId: PropTypes.number,
        dispatch: PropTypes.func,
        subtypesLoading: PropTypes.bool
    };

    constructor(props) {
        super(props);
    }

    componentDidMount() {
        this.props.dispatch(loadPublicationSubtypesList(this.props.vocabId));
    }

    render() {
        return (
            <div>
                {
                    this.props.subtypesLoading &&
                    <SelectField fullWidth className="requiredField">
                        <MenuItem disabled />
                    </SelectField>
                }
                {
                    !this.props.subtypesLoading &&
                    <PublicationSubtypeForm { ...this.props } />
                }
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        subtypes: state.get('publicationSubtypesReducer').subtypes || [],
        subtypesLoading: state.get('publicationSubtypesReducer').subtypesLoading || false
    };
};

export default connect(mapStateToProps)(PublicationSubtypeFormContainer);
