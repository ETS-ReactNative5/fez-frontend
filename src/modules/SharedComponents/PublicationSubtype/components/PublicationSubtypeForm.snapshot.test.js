jest.dontMock('./PublicationSubtypeForm');

import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { PublicationSubtypeForm } from './PublicationSubtypeForm';

function setup(props) {
    return shallow(<PublicationSubtypeForm {...props} />);
}

describe('PublicationSubtypeForm snapshots tests', () => {
    it('renders correctly', () => {
        const props = {
            vocabId: 453581,
            subtypesList: []
        };

        const wrapper = setup(props);

        const tree = toJson(wrapper);

        expect(tree).toMatchSnapshot();
    });

    it('renders correctly 2', () => {
        const props = {
            vocabId: 453581,
            subtypesList: [
                {
                    'cvr_id': 4706,
                    'cvr_parent_cvo_id': 453581,
                    'cvr_child_cvo_id': 453582,
                    'controlled_vocab': {
                        'cvo_id': 453582,
                        'cvo_title': 'Research book (original research)',
                        'cvo_desc': '',
                        'cvo_image_filename': null,
                        'cvo_external_id': null,
                        'cvo_hide': 0,
                        'cvo_order': 2,
                        'cvo_lat': null,
                        'cvo_long': null,
                        'cvo_policy': null,
                        'controlled_vocab_children': []
                    }
                }, {
                    'cvr_id': 4707,
                    'cvr_parent_cvo_id': 453581,
                    'cvr_child_cvo_id': 453583,
                    'controlled_vocab': {
                        'cvo_id': 453583,
                        'cvo_title': 'Textbook',
                        'cvo_desc': '',
                        'cvo_image_filename': null,
                        'cvo_external_id': null,
                        'cvo_hide': 0,
                        'cvo_order': 4,
                        'cvo_lat': null,
                        'cvo_long': null,
                        'cvo_policy': null,
                        'controlled_vocab_children': []
                    }
                }, {
                    'cvr_id': 4708,
                    'cvr_parent_cvo_id': 453581,
                    'cvr_child_cvo_id': 453584,
                    'controlled_vocab': {
                        'cvo_id': 453584,
                        'cvo_title': 'Edited book',
                        'cvo_desc': '',
                        'cvo_image_filename': null,
                        'cvo_external_id': null,
                        'cvo_hide': 0,
                        'cvo_order': 6,
                        'cvo_lat': null,
                        'cvo_long': null,
                        'cvo_policy': null,
                        'controlled_vocab_children': []
                    }
                }, {
                    'cvr_id': 4709,
                    'cvr_parent_cvo_id': 453581,
                    'cvr_child_cvo_id': 453585,
                    'controlled_vocab': {
                        'cvo_id': 453585,
                        'cvo_title': 'Reference work, encyclopaedia, manual or handbook',
                        'cvo_desc': '',
                        'cvo_image_filename': null,
                        'cvo_external_id': null,
                        'cvo_hide': 0,
                        'cvo_order': 8,
                        'cvo_lat': null,
                        'cvo_long': null,
                        'cvo_policy': null,
                        'controlled_vocab_children': []
                    }
                }, {
                    'cvr_id': 4710,
                    'cvr_parent_cvo_id': 453581,
                    'cvr_child_cvo_id': 453586,
                    'controlled_vocab': {
                        'cvo_id': 453586,
                        'cvo_title': 'Creative work',
                        'cvo_desc': '',
                        'cvo_image_filename': null,
                        'cvo_external_id': null,
                        'cvo_hide': 0,
                        'cvo_order': 10,
                        'cvo_lat': null,
                        'cvo_long': null,
                        'cvo_policy': null,
                        'controlled_vocab_children': []
                    }
                }, {
                    'cvr_id': 4711,
                    'cvr_parent_cvo_id': 453581,
                    'cvr_child_cvo_id': 453587,
                    'controlled_vocab': {
                        'cvo_id': 453587,
                        'cvo_title': 'Other',
                        'cvo_desc': '',
                        'cvo_image_filename': null,
                        'cvo_external_id': null,
                        'cvo_hide': 0,
                        'cvo_order': 12,
                        'cvo_lat': null,
                        'cvo_long': null,
                        'cvo_policy': null,
                        'controlled_vocab_children': []
                    }
                }
            ]
        };

        const wrapper = setup(props);

        const tree = toJson(wrapper);

        expect(tree).toMatchSnapshot();
    });
});
