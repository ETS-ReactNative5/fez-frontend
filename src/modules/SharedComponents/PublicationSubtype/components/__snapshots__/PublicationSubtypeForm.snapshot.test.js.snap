// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PublicationSubtypeForm snapshots tests renders correctly 1`] = `
<SelectField
  autoWidth={false}
  disabled={false}
  floatingLabelText="Publication subtype"
  fullWidth={true}
  maxHeight={250}
  multiple={false}
  name="selectedValue"
  onChange={[Function]}
  value={null}
>
  <MenuItem
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
    checked={false}
    desktop={false}
    disabled={true}
    focusState="none"
    insetChildren={false}
    primaryText="Publication subtype"
  />
</SelectField>
`;

exports[`PublicationSubtypeForm snapshots tests renders correctly 2 1`] = `
<SelectField
  autoWidth={false}
  disabled={false}
  floatingLabelText="Publication subtype"
  fullWidth={true}
  maxHeight={250}
  multiple={false}
  name="selectedValue"
  onChange={[Function]}
  value={null}
>
  <MenuItem
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
    checked={false}
    desktop={false}
    disabled={true}
    focusState="none"
    insetChildren={false}
    primaryText="Publication subtype"
  />
  <MenuItem
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
    checked={false}
    desktop={false}
    disabled={false}
    focusState="none"
    insetChildren={false}
    primaryText="Research book (original research)"
    value="Research book (original research)"
  />
  <MenuItem
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
    checked={false}
    desktop={false}
    disabled={false}
    focusState="none"
    insetChildren={false}
    primaryText="Textbook"
    value="Textbook"
  />
  <MenuItem
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
    checked={false}
    desktop={false}
    disabled={false}
    focusState="none"
    insetChildren={false}
    primaryText="Edited book"
    value="Edited book"
  />
  <MenuItem
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
    checked={false}
    desktop={false}
    disabled={false}
    focusState="none"
    insetChildren={false}
    primaryText="Reference work, encyclopaedia, manual or handbook"
    value="Reference work, encyclopaedia, manual or handbook"
  />
  <MenuItem
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
    checked={false}
    desktop={false}
    disabled={false}
    focusState="none"
    insetChildren={false}
    primaryText="Creative work"
    value="Creative work"
  />
  <MenuItem
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
    checked={false}
    desktop={false}
    disabled={false}
    focusState="none"
    insetChildren={false}
    primaryText="Other"
    value="Other"
  />
</SelectField>
`;
